File: Ashutosh-Singh/SlidingWindow/Minimum Operations-to-Reduce-X-to-Zero.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int minOperations(vector<int>& nums, int x) {
        int sum = -x;
        for (int i = 0; i < nums.size(); i++) {
            sum += nums[i];
        }
        if (sum < 0) return -1;

        if (sum == 0) return nums.size();
        int st = 0, end = 0, maxLength = 0;

        while (end < nums.size()) {
            sum -= nums[end];

            while (sum < 0) {
                sum += nums[st++];
            }

            if (sum == 0) {
                maxLength = max(maxLength, end - st + 1);
            }
            end++;
        }
        return maxLength == 0 ? -1 : nums.size() - maxLength;
    }
};

int main() {
    int n, x;
    cin >> n >> x;

    vector<int> nums(n);
    for (int i = 0; i < n; ++i) {
        cin >> nums[i];
    }

    Solution solution;
    cout << solution.minOperations(nums, x) << endl;

    return 0;
}

File: Anushka-Pachauri/SlidingWindow/MinimumOperationsToReduceXToZero.java
class Solution {
    public int minOperations(int[] nums, int x) {
        int totalSum=0;
        for(int num : nums){
            totalSum+=num;
        }
        if(x>totalSum){
            return -1;
        }
        if(x==totalSum){
            return nums.length;
        }
        int target=totalSum-x;
        int start=0;
        int end=0;
        int maxLength=Integer.MIN_VALUE;
        int sum=0;
        while(end<nums.length){
            sum+=nums[end];
            while(sum>target && start<=end){
                sum-=nums[start];
                start++;
            }
            if(sum==target){
                maxLength=Math.max(maxLength,end-start+1);
            }
            end++;
        }
        return maxLength==Integer.MIN_VALUE ? -1 : nums.length-maxLength;
    }
}